### DISCUSSION TITLE
How to integrate pythonocc-core into a Blender addon

### DISCUSSION THREAD

**Discussion by marvin-kluge** [IfcOpenShell/IfcOpenShell#3507] [Q&A] [answered]
This is not so much an issue as a question, but I hope to find an answer in this community.

We are working on a research project and want to create a Blender addon built on top of BlenderBIM. Our python library uses IfcOpenShell and PythonOCC. The problem is that current versions of PythonOCC can only be installed via Anaconda, and we can't manage to integrate PythonOCC with Blender's built-in Python interpreter.
I saw, that you used PythonOCC a couple years ago in BlenderBIM as well and i was hoping, that you maybe could provide some insight how we could manage to implement it in a blender addon as well.

The error we are getting is:
`ImportError: DLL load failed while importing _BRep: The specified module was not found.` when PythonOCC tries to import a .pyd file.

We have tried:

- copy a complete anaconda installation of PythonOCC including all dependencies into the BlenderBIM interpreter.
- Build PythonOCC ourselves, which failed (unfortunately the instructions in the readme.md seem to be out of date).
- deliver all packages including PythonOCC with the addon to blender and add the packages folder to the sys.path (much like you did it in BlenderBIM with the site/packages folder)

Does anyone have any ideas on how to integrate PythonOCC correctly, or can direct me to the right place to look for more information? Any help is appreciated.

**Comment by Moult** [comment]
It was definitely tricky to do back in the day, but I reckon I can get it working again. Basically some dependencies come straight from Conda (e.g. https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/Makefile#L55 ) so the key is to work out what the dependencies from Conda are (pretty simple, just install Conda and see what it pulls in and make a list) then the second trick is to make sure all the paths are correct and the RPATHS are appropriate for the dlls and whatnot. Conda uses an RPATH 3 of `../../` which is why our libs dir is 3 levels deep: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/__init__.py#L35-L38

So it sounds like your option 3.

Happy to screenshare and get it working with you. I'm available 9 to 5pm Sydney time.

**Reply by marvin-kluge** [reply]
Thank you for your quick answer.  I will try this as soon as i can but unfortunately another project got in the way and I will contact you in the future if thats okay with you.
Thanks for the help.

**Reply by marvin-kluge** [reply]
Hey @Moult, I had now time to try your suggestions and unfortunately, I got the same error as before.
I added all packages that conda installs when installing pythonocc into the site-packages folder inside our blender plugin and used the same folder structure you use in your plugin (libs/site/packages).
Then, in the register() function, I tried to import a submodule from OCC.Core (such as gp or BRep). I have to add here that I can import occ per se but not the submodules.
I noticed that you had a comment above the function in which you imported pythonocc when you were still using it which said, that you had to import it as late as possible. Maybe it has something to do with this?
Thanks for the help.

**Reply by aothms** [reply, marked as answer]
> I have to add here that I can import occ per se but not the submodules.

That is indeed possible. `import OCC` itself does not actually look for the binaries yet.

What OS are you on?

For windows there is https://github.com/lucasg/Dependencies which you could try to open e.g _BRep.pyd and see if it reports on any missing dll dependencies.

On linux it might help to run in `strace` and grep on things like .so or .pyd to see the various attempts of libraries being loaded.

**Reply by marvin-kluge** [reply]
Thanks, this dependencies tool really helped a lot. I got pooc running

**Reply by marvin-kluge** [reply]
Sorry for asking again but I got the same error for _V3d now. It seems like this is the only module i cannot import from OCC. I tried the same approach as before with importing _V3d into the dependencies tool and providing the necessary DLLs with the blender addon but it seems like this doesn't work for _V3d.
Has any of you maybe a suggestion as to why this happened?

**Reply by aothms** [reply]
I guess _V3d depends on OpenGL, I don't know if updating graphics drivers might help (I think the OpenGL dll comes with Windows itself, not the drivers, but not entirely sure). Could also be that since Blender also depends on OpenGL but maybe uses mesa or some other implementation that simply the OCCT libs are not compatible with Blender.

Do you really need _V3d though. Isn't it a better experience if you try and use the Blender viewport to render?

**Reply by marvin-kluge** [reply]
_V3d is needed for IfcOpenshell.geom and that is used in the tool for which I'm implementing the blender addon for. 
Maybe it is just not possible to develop a blender addon for this specific tool.
Thanks for the help

**Reply by aothms** [reply]
I think V3d is only in the occ_utils initialize_display() https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/geom/occ_utils.py#L65 I don't know of course what you're exactly trying to do, but unless you really depend on this embedded pyocc viewer it should be possible to not import it. I don't mind if you propose some changes to ifcopenshell.geom so that it imports more selectively,
